pkg/solana/chainreader/event_read_binding.go

```diff
@@ -5,54 +5,142 @@ import (
 )
 
 type eventReadBinding struct {
+	// dependencies
+	reader   EventsReader
+	remapper remapHelper
+	codec    types.RemoteCodec
+	modifier commoncodec.Modifier
+	conf     config.PollingFilter
+	// filter in eventReadBinding is to be used as an override for lp filter defined in the namespace binding.
+	// If filter is nil, this event should be registered with the lp filter defined in the namespace binding.
+	filter *syncedFilter
+
+	// static data
 	namespace, genericName string
-	codec                  types.RemoteCodec
-	modifier               commoncodec.Modifier
-	key                    solana.PublicKey
-	remapper               remapHelper
-	indexedSubKeys         map[string]uint64
-	reader                 EventsReader
 	eventSig               [logpoller.EventSignatureLength]byte
+	indexedSubKeys         *indexedSubkeys
 	readDefinition         config.ReadDefinition
+
+	// thread protected fields
+	mu             sync.RWMutex
+	key            solana.PublicKey
+	bound          bool
+	registerCalled bool
 }
 
 func newEventReadBinding(
 	namespace, genericName string,
-	indexedSubKeys map[string]uint64,
+	indexedSubKeys *indexedSubkeys,
 	reader EventsReader,
-	eventSig [logpoller.EventSignatureLength]byte,
 	readDefinition config.ReadDefinition,
+	conf config.PollingFilter,
 ) *eventReadBinding {
 	binding := &eventReadBinding{
+		filter:         newSyncedFilter(),
 		namespace:      namespace,
 		genericName:    genericName,
 		indexedSubKeys: indexedSubKeys,
 		reader:         reader,
-		eventSig:       eventSig,
 		readDefinition: readDefinition,
+		conf:           conf,
 	}
 
 	binding.remapper = remapHelper{binding.remapPrimitive}
 
 	return binding
 }
 
+func (b *eventReadBinding) Bind(ctx context.Context, address solana.PublicKey) error {
+	if b.isBound() {
+		// we are changing contract address reference, so we need to unregister old filter if it exists
+		if err := b.Unregister(ctx); err != nil {
+			return err
+		}
+	}
+
+	// filter isn't required here because the event can also be polled for by the contractBinding common filter.
+	if b.filter != nil {
+		b.filter.SetName(fmt.Sprintf("%s.%s.%s", b.namespace, b.genericName, uuid.NewString()))
+		b.filter.SetAddress(address)
+	}
+
+	b.setBinding(address)
+
+	if b.registered() {
+		return b.Register(ctx)
+	}
+
+	return nil
+}
+
+func (b *eventReadBinding) Unbind(ctx context.Context) error {
+	if !b.isBound() {
+		return nil
+	}
+
+	if b.filter != nil {
+		b.filter.SetAddress(solana.PublicKey{})
+		b.filter.SetName("")
+	}
+
+	b.unsetBinding()
+
+	return b.Unregister(ctx)
+}
+
+func (b *eventReadBinding) Register(ctx context.Context) error {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	if b.filter == nil {
+		return nil
+	}
+
+	b.registerCalled = true
+
+	// can't be true before filters params are set so there is no race with a bad filter outcome
+	if !b.bound {
+		return nil
+	}
+
+	return b.filter.Register(ctx, b.reader)
```

Comment:
Do we still want to set `b.registerCalled = true` if we return before calling `b.filter.Register(ctx, b.reader)`?
```suggestion
	// can't be true before filters params are set so there is no race with a bad filter outcome
	if !b.bound {
		return nil
	}
	b.registerCalled = true
	return b.filter.Register(ctx, b.reader)
```

Reply:
It's the same in EVM. I'm not quite sure why, but I think a lot of this on that side was done by @ilija42.